lib/srv/db/opensearch/engine.go

```diff
@@ -0,0 +1,283 @@
+	signingSvc *libaws.SigningService
+
+	// EngineConfig is the common database engine configuration.
+	common.EngineConfig
+	// clientConn is a client connection.
+	clientConn net.Conn
+	// sessionCtx is current session context.
+	sessionCtx *common.Session
+
+	// for tests
+	GetSigningCredsFn libaws.GetSigningCredentialsFunc
+}
+
+// InitializeConnection initializes the engine with the client connection.
+func (e *Engine) InitializeConnection(clientConn net.Conn, sessionCtx *common.Session) error {
+	e.clientConn = clientConn
+	e.sessionCtx = sessionCtx
+
+	awsSession, err := e.CloudClients.GetAWSSession(sessionCtx.Database.GetAWS().Region)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	svc, err := libaws.NewSigningService(libaws.SigningServiceConfig{
+		Clock:                 e.Clock,
+		Session:               awsSession,
+		GetSigningCredentials: e.GetSigningCredsFn,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	e.signingSvc = svc
+
+	return nil
+}
+
+// SendError sends an error to OpenSearch client.
+func (e *Engine) SendError(err error) {
+	if e.clientConn == nil || err == nil || utils.IsOKNetworkError(err) {
+		return
+	}
+
+	// TODO: ideally we would to switch over to strongly typed opensearch errors, but there is no equivalent of elastic.ErrorCause in opensearch libraries.
+	reason := err.Error()
+	cause := elastic.ErrorCause{
+		Reason: &reason,
+		Type:   "internal_server_error_exception",
+	}
+
+	// Assume internal server error HTTP 500 and override if possible.
+	statusCode := http.StatusInternalServerError
+	if trace.IsAccessDenied(err) {
+		statusCode = http.StatusUnauthorized
+		cause.Type = "access_denied_exception"
+	}
+
+	jsonBody, err := json.Marshal(cause)
+	if err != nil {
+		e.Log.WithError(err).Error("failed to marshal error response")
+		return
+	}
+
+	response := &http.Response{
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		StatusCode: statusCode,
+		Body:       io.NopCloser(bytes.NewBuffer(jsonBody)),
+		Header: map[string][]string{
+			"Content-Type":   {"application/json"},
+			"Content-Length": {strconv.Itoa(len(jsonBody))},
+		},
+	}
+
+	if err := response.Write(e.clientConn); err != nil {
+		e.Log.Errorf("OpenSearch error: %+v", trace.Unwrap(err))
+		return
+	}
+}
+
+// HandleConnection authorizes the incoming client connection, connects to the
+// target OpenSearch server and starts proxying requests between client/server.
+func (e *Engine) HandleConnection(ctx context.Context, _ *common.Session) error {
+	err := e.checkAccess(ctx)
+	e.Audit.OnSessionStart(e.Context, e.sessionCtx, err)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer e.Audit.OnSessionEnd(e.Context, e.sessionCtx)
+
+	clientConnReader := bufio.NewReader(e.clientConn)
+	for {
+		req, err := http.ReadRequest(clientConnReader)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if err := e.process(ctx, req); err != nil {
+			return trace.Wrap(err)
+		}
+	}
```

Comment:
During testing I'm getting setup on blocking calls: 
```
curl -XGET  'http://localhost:52201/movies/_search' \
   -H 'Content-Type: application/json'  \
   -d '
{
  "query": {
    "match_all": {}
  }
}
'
```
In my case this HTTP call blacks the client forever.

Reply:
I'll take a look.

Reply:
Running `curl -v` prints a warning: 
`* no chunk, no close, no size. Assume close to signal end`

I chose to calculate the content length (if missing), and now `curl` closes the connection promptly.